File: android/app/src/main/kotlin/com/elipair/copsandrobbers/CommunityNativeAdFactory.kt
package com.elipair.copsandrobbers

import android.content.Context
import android.graphics.Typeface
import android.graphics.drawable.GradientDrawable
import android.text.TextUtils
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.widget.ImageView
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.TextView
import com.google.android.gms.ads.nativead.MediaView
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdView
import io.flutter.FlutterInjector
import io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin.NativeAdFactory

class CommunityNativeAdFactory(private val context: Context) : NativeAdFactory {
    private val fonts = listOf("Medium", "SemiBold").associateWith {
        Typeface.createFromAsset(context.assets, FlutterInjector.instance().flutterLoader()
            .getLookupKeyForAsset("assets/fonts/Pretendard-$it.ttf"))
    }

    private fun dp(value: Number) = (value.toFloat() * context.resources.displayMetrics.density).toInt()

    override fun createNativeAd(nativeAd: NativeAd, customOptions: MutableMap<String, Any>?): NativeAdView {
        val options = requireNotNull(customOptions)
        fun number(key: String) = (options.getValue(key) as Number)
        fun label(value: String?, size: String, color: String, lines: Int = 1, bold: Boolean = false) =
            TextView(context).apply {
                text = value
                setTextSize(TypedValue.COMPLEX_UNIT_DIP, number(size).toFloat())
                setTextColor(number(color).toInt())
                typeface = fonts[if (bold) "SemiBold" else "Medium"]
                maxLines = lines
                ellipsize = TextUtils.TruncateAt.END
                includeFontPadding = false
                visibility = if (value.isNullOrEmpty()) View.GONE else View.VISIBLE
            }

        val adView = NativeAdView(context)
        val content = LinearLayout(context).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER_VERTICAL
            setPadding(dp(22), dp(1), dp(22), dp(1))
        }
        adView.addView(content, FrameLayout.LayoutParams(-1, -1))
        val copy = LinearLayout(context).apply { orientation = LinearLayout.VERTICAL }
        content.addView(copy, LinearLayout.LayoutParams(0, -2, 1f).apply { rightMargin = dp(12) })
        val header = LinearLayout(context).apply {
            gravity = Gravity.CENTER_VERTICAL
        }
        copy.addView(header, LinearLayout.LayoutParams(-1, -2))
        val badge = label(options.getValue("adLabel") as String, "captionSize", "secondaryColor").apply {
            setPadding(dp(4), dp(2), dp(4), dp(2))
            background = GradientDrawable().apply {
                setColor(number("badgeColor").toInt())
                cornerRadius = dp(4).toFloat()
            }
        }
        header.addView(badge)
        val headline = label(nativeAd.headline, "headlineSize", "textColor", bold = true)
        header.addView(headline, LinearLayout.LayoutParams(0, -2, 1f).apply { leftMargin = dp(8) })
        adView.headlineView = headline
        val body = label(nativeAd.body, "bodySize", "secondaryColor")
        copy.addView(body, LinearLayout.LayoutParams(-1, -2).apply { topMargin = dp(4) })
        adView.bodyView = body
        val footer = LinearLayout(context).apply { gravity = Gravity.CENTER_VERTICAL }
        copy.addView(footer, LinearLayout.LayoutParams(-1, -2).apply { topMargin = dp(4) })
        val icon = ImageView(context).apply {
            setImageDrawable(nativeAd.icon?.drawable)
            scaleType = ImageView.ScaleType.FIT_CENTER
            visibility = if (nativeAd.icon == null) View.GONE else View.VISIBLE
        }
        footer.addView(icon, LinearLayout.LayoutParams(dp(24), dp(24)))
        adView.iconView = icon
        val advertiser = label(nativeAd.advertiser, "captionSize", "secondaryColor")
        footer.addView(advertiser, LinearLayout.LayoutParams(0, -2, 1f).apply { leftMargin = dp(8) })
        adView.advertiserView = advertiser

        val action = label(nativeAd.callToAction, "bodySize", "accentColor", bold = true)
        footer.addView(action, LinearLayout.LayoutParams(0, -2, 1f).apply { leftMargin = dp(8) })
        adView.callToActionView = action
        val media = MediaView(context).apply {
            mediaContent = nativeAd.mediaContent
            setImageScaleType(ImageView.ScaleType.CENTER_INSIDE)
        }
        content.addView(media, LinearLayout.LayoutParams(dp(120), dp(120)))
        adView.mediaView = media
        adView.setNativeAd(nativeAd)
        return adView
    }
}

File: ios/Runner/CommunityNativeAdFactory.swift
import CoreText
import Flutter
import GoogleMobileAds
import UIKit
import google_mobile_ads

final class CommunityNativeAdFactory: NSObject, FLTNativeAdFactory {
  private static let registerFonts: Void = {
    for weight in ["Medium", "SemiBold"] {
      let key = FlutterDartProject.lookupKey(forAsset: "assets/fonts/Pretendard-\(weight).ttf")
      if let path = Bundle.main.path(forResource: key, ofType: nil) {
        CTFontManagerRegisterFontsForURL(URL(fileURLWithPath: path) as CFURL, .process, nil)
      }
    }
  }()

  func createNativeAd(_ nativeAd: NativeAd, customOptions: [AnyHashable: Any]? = nil) -> NativeAdView? {
    guard let options = customOptions else { return nil }
    _ = Self.registerFonts
    func number(_ key: String) -> CGFloat { CGFloat((options[key] as? NSNumber)?.doubleValue ?? 0) }
    func color(_ key: String) -> UIColor {
      let value = (options[key] as? NSNumber)?.uint32Value ?? 0
      return UIColor(red: CGFloat((value >> 16) & 255) / 255,
                     green: CGFloat((value >> 8) & 255) / 255,
                     blue: CGFloat(value & 255) / 255, alpha: CGFloat(value >> 24) / 255)
    }
    func label(_ text: String?, size: String, colorKey: String, lines: Int = 1, bold: Bool = false) -> UILabel {
      let label = UILabel()
      label.text = text
      label.font = UIFont(name: bold ? "Pretendard-SemiBold" : "Pretendard-Medium", size: number(size))
        ?? UIFont.systemFont(ofSize: number(size), weight: bold ? .semibold : .medium)
      label.textColor = color(colorKey)
      label.numberOfLines = lines
      label.lineBreakMode = .byTruncatingTail
      label.isHidden = text?.isEmpty ?? true
      return label
    }

    let adView = NativeAdView()
    let badgeText = label(options["adLabel"] as? String, size: "captionSize", colorKey: "secondaryColor")
    let badge = UIView()
    badge.backgroundColor = color("badgeColor")
    badge.layer.cornerRadius = 4
    badgeText.translatesAutoresizingMaskIntoConstraints = false
    badge.addSubview(badgeText)
    NSLayoutConstraint.activate([
      badgeText.leadingAnchor.constraint(equalTo: badge.leadingAnchor, constant: 4),
      badgeText.trailingAnchor.constraint(equalTo: badge.trailingAnchor, constant: -4),
      badgeText.topAnchor.constraint(equalTo: badge.topAnchor, constant: 2),
      badgeText.bottomAnchor.constraint(equalTo: badge.bottomAnchor, constant: -2),
    ])
    let icon = UIImageView(image: nativeAd.icon?.image)
    icon.contentMode = .scaleAspectFit
    icon.isHidden = nativeAd.icon == nil
    icon.widthAnchor.constraint(equalToConstant: 24).isActive = true
    icon.heightAnchor.constraint(equalToConstant: 24).isActive = true
    adView.iconView = icon
    let advertiser = label(nativeAd.advertiser, size: "captionSize", colorKey: "secondaryColor")
    adView.advertiserView = advertiser
    advertiser.setContentCompressionResistancePriority(.defaultLow, for: .horizontal)
    let header = UIStackView(arrangedSubviews: [badge, icon, advertiser, UIView()])
    header.spacing = 8
    header.alignment = .center
    header.isLayoutMarginsRelativeArrangement = true
    header.layoutMargins = UIEdgeInsets(top: 0, left: 0, bottom: 0, right: 20)
    header.heightAnchor.constraint(equalToConstant: max(24, number("captionSize") * 1.4)).isActive = true

    let headline = label(nativeAd.headline, size: "headlineSize", colorKey: "textColor", lines: 2, bold: true)
    adView.headlineView = headline
    let body = label(nativeAd.body, size: "bodySize", colorKey: "secondaryColor", lines: 2)
    adView.bodyView = body
    let action = label(nativeAd.callToAction, size: "bodySize", colorKey: "accentColor", bold: true)
    // 클릭 처리는 SDK가 맡는다. Flutter의 게시글 탭과 공유하지 않는다.
    action.isUserInteractionEnabled = false
    adView.callToActionView = action
    let copy = UIStackView(arrangedSubviews: [headline, body, action])
    copy.axis = .vertical
    copy.spacing = 8
    copy.setCustomSpacing(12, after: body)
    let media = MediaView()
    media.mediaContent = nativeAd.mediaContent
    media.contentMode = .scaleAspectFit
    media.widthAnchor.constraint(equalToConstant: 120).isActive = true
    media.heightAnchor.constraint(equalToConstant: 120).isActive = true
    adView.mediaView = media
    let row = UIStackView(arrangedSubviews: [copy, media])
    row.spacing = 12
    row.alignment = .center
    let content = UIStackView(arrangedSubviews: [header, row])
    content.axis = .vertical
    content.spacing = 8
    content.translatesAutoresizingMaskIntoConstraints = false
    adView.addSubview(content)
    NSLayoutConstraint.activate([
      content.leadingAnchor.constraint(equalTo: adView.leadingAnchor, constant: 22),
      content.trailingAnchor.constraint(equalTo: adView.trailingAnchor, constant: -22),
      content.topAnchor.constraint(equalTo: adView.topAnchor, constant: 16),
      content.bottomAnchor.constraint(equalTo: adView.bottomAnchor, constant: -16),
    ])
    adView.nativeAd = nativeAd
    return adView
  }
}
